numpy/core/arrayprint.py

```diff
@@ -541,23 +554,31 @@ def array2string(a, max_line_width=None, precision=None,
     '[0x0L 0x1L 0x2L]'
 
     """
-    # Deprecation 05-16-2017  v1.14
-    if style is not np._NoValue:
-        warnings.warn("'style' argument is deprecated and no longer functional",
-                      DeprecationWarning, stacklevel=3)
+    legacy = kwarg.pop('legacy', None)
+    if kwarg:
+        msg = "array2string() got unexpected keyword argument '{}'"
+        raise TypeError(msg.format(kwarg.popitem()[0]))
 
     overrides = _make_options_dict(precision, threshold, edgeitems,
                                    max_line_width, suppress_small, None, None,
-                                   sign, formatter, floatmode)
+                                   sign, formatter, floatmode, legacy)
     options = _format_options.copy()
     options.update(overrides)
 
+    if options['legacy'] and a.shape == () and not a.dtype.names:
+        return style(a.item())
+    elif not (style is np._NoValue):
+        warnings.warn("'style' argument is deprecated and no longer functional"
+                      " except in 'legacy' mode",
+                      DeprecationWarning, stacklevel=3)
```

Comment:
This logic isn't right, it will warn for `array2str([1], style=str, legacy=True)` - you need an outer `if` switching on `legacy`.